### PR TITLE
feat(theming): Only convert a background image if it benefits from it

### LINKED ISSUE #34233: [Bug]: Uploading Theme Login Image inexplicably increases image size significantly
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Through Settings > Theming I uploaded an jpg as the login background. 

The Image I uploaded was 1079x609 and 47KB in size (I used an unrealistically small image for this example). After the upload completed I logged out and refreshed the login page to find that the correct image is appearing with the same dimensions for the background image but its now 23x larger in size (1.1MB). I tried again with a larger image of 2k at 200KB (what I would consider a realistic size for a background image) and the resulting image to download is 3.4MB which can take up to 30+ seconds to load on slower mobile networks (I aim for my background image to be 100KB-300KB depending on content).

I cleared any browser caching and server caching (I use redis) to ensure the correct image is being delivered but I get the same result.

I could see Nextcloud modifying the image to optimize it after upload but I don't understand why the image size should increase so dramatically (especially without any perceivable difference in the final background image).

Screen Shot 2022-09-23 at 9 10 04 PM

I'm using Nextcloud in the official docker container on Unraid 6.11.0 (Linux 5.19.9-Unraid x86_64)
The browser I am using is Chrome.

### Steps to reproduce

1. Go to Settings > Theming > Login image and upload a ~200kb jpg 
2. After upload saves view the either the preview image in dev tools or logout and view the login background in dev tools
3. See that while the image is the correct image the file size is significantly larger (~3MB) than what was uploaded.

### Expected behavior

I expect the theming login image to remain the same size or at least be smaller (if nextcloud optimizes the image after upload).

### Installation method

Community Docker image

### Operating system

Other

### PHP engine version

PHP 7.4

### Web server

Nginx

### Database engine version

PostgreSQL

### Is this bug present after an update or on a fresh install?

_No response_

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "default_phone_region": "us",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "filesystem_check_changes": 1,
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "24.0.3.2",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "5432",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "app_install_overwrite": [
            "documentserver_community"
        ],
        "maintenance": false,
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "ssl",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.10.0
  - activity: 2.16.0
  - bruteforcesettings: 2.4.0
  - circles: 24.0.0
  - cloud_federation_api: 1.7.0
  - comments: 1.14.0
  - contactsinteraction: 1.5.0
  - dashboard: 7.4.0
  - dav: 1.22.0
  - federatedfilesharing: 1.14.0
  - files: 1.19.0
  - files_pdfviewer: 2.5.0
  - files_rightclick: 1.3.0
  - files_sharing: 1.16.2
  - files_trashbin: 1.14.0
  - files_versions: 1.17.0
  - files_videoplayer: 1.13.0
  - logreader: 2.9.0
  - lookup_server_connector: 1.12.0
  - nextcloud_announcements: 1.13.0
  - notifications: 2.12.0
  - oauth2: 1.12.0
  - onlyoffice: 7.5.4
  - password_policy: 1.14.0
  - photos: 1.6.0
  - privacy: 1.8.0
  - provisioning_api: 1.14.0
  - recommendations: 1.3.0
  - serverinfo: 1.14.0
  - settings: 1.6.0
  - sharebymail: 1.14.0
  - support: 1.7.0
  - survey_client: 1.12.0
  - systemtags: 1.14.0
  - text: 3.5.1
  - theming: 1.15.0
  - twofactor_backupcodes: 1.13.0
  - updatenotification: 1.14.0
  - user_status: 1.4.0
  - viewer: 1.8.0
  - weather_status: 1.4.0
  - workflowengine: 2.6.0
Disabled:
  - admin_audit
  - documentserver_community: 0.1.12
  - encryption
  - federation: 1.14.0
  - files_external
  - firstrunwizard: 2.13.0
  - user_ldap
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

I'm using Nextcloud in the official docker container on Unraid 6.11.0 (Linux 5.19.9-Unraid x86_64)
The browser I am using is Chrome.

### PR DESCRIPTION
**Pull request by susnux**
* Resolves: #34233

## Summary
* WebP images are generally quite small, converting to pngs would increase the filesize a lot.
* Small JPEG and PNG images do not benefit from any conversion, so skip it.
* JPEG images will get quite bigger when converted to PNG so instead convert to progressive JPEG

This also fixes the wrong mime type sent to the browser, as before this every non gif, non svg would have been converted to a png. But the server sent the file extension as the mime type. Now the file type is preserved and only converted to the same file type.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/theming/lib/ImageManager.php [hunk 1]
```diff
@@ -243,32 +243,79 @@ public function updateImage(string $key, string $tmpFile): string {
...
+	/**
+	 * Decide whether an image benefits from shrinking and reconverting
+	 *
+	 * @param string $mimeType the mime type of the image
+	 * @param int $contentSize size of the image file
+	 * @return bool
+	 */
+	private function shouldOptimizeBackgroundImage(string $mimeType, int $contentSize): bool {
+		// Do not touch SVGs
+		if (strpos($mimeType, 'image/svg') !== false) {
```
Comment (szaimen) [inline apps/theming/lib/ImageManager.php:303]: are you sure with that?
https://github.com/nextcloud/server/blob/7858b8313ab098f8d0f7c0783f6051215d02eb56/resources/config/mimetypemapping.dist.json#L180

**Review by juliusknorr — APPROVED** [review]

**Comment by szaimen** [comment]
@skjnldsv therr is also this attempt since you replied in https://github.com/nextcloud/server/pull/32787

**Review by skjnldsv — APPROVED** [review]

**Comment by Ornanovitch** [comment]
Hey @skjnldsv, could this be backported to 25? That would be very useful. Thanks!

**Pull request merged**